qiskit/circuit/qpy_serialization.py

```diff
@@ -954,25 +949,62 @@ def _read_circuit(file_obj):
         num_registers,
         num_instructions,
     ) = header
-    registers = {}
+    out_registers = {"q": {}, "c": {}}
     if num_registers > 0:
         circ = QuantumCircuit(name=name, global_phase=global_phase, metadata=metadata)
-        # TODO Update to handle registers composed of not continuous bit
-        # indices. Right now this only works for standalone registers or
-        # registers composed bit indices that are continuous
         registers = _read_registers(file_obj, num_registers)
-        for qreg in registers["q"].values():
-            min_index = min(qreg["index_map"].keys())
-            qubits = [Qubit() for i in range(min_index - len(circ.qubits))]
-            if qubits:
+        standalone_qreg = {key: value[1] for key, value in registers["q"].items() if value[0]}
+        standalone_creg = {key: value[1] for key, value in registers["c"].items() if value[0]}
+        shared_qreg = {key: value[1] for key, value in registers["q"].items() if not value[0]}
+        shared_creg = {key: value[1] for key, value in registers["c"].items() if not value[0]}
+        standalone_qreg_starts = {
+            name: min(value)
+            for name, value in sorted(standalone_qreg.items(), key=lambda v: min(v[1]))
+        }
+        standalone_creg_starts = {
+            name: min(value)
+            for name, value in sorted(standalone_creg.items(), key=lambda v: min(v[1]))
+        }
+        qubit_count = 0
+        # Add quantum registers and bits
+        for name, start in standalone_qreg_starts.items():
+            if start > qubit_count:
+                qubits = [Qubit() for _ in range(start - qubit_count)]
                 circ.add_bits(qubits)
-            circ.add_register(qreg["register"])
-        for creg in registers["c"].values():
-            min_index = min(creg["index_map"].keys())
-            clbits = [Clbit() for i in range(min_index - len(circ.clbits))]
-            if clbits:
+            qubit_indices = standalone_qreg[name]
+            num_reg_bits = len(qubit_indices)
+            qreg = QuantumRegister(num_reg_bits, name)
+            circ.add_register(qreg)
```

Comment:
If the `Bit`s in the `Register` are non-contiguous, could this still lead to incorrectly reconstructing the circuit `Bit` order? Since the `Bit`s in `qreg` will be added to `circ.qubits` in order, it looks like their indices `qubit_indices` will be lost.

Reply:
Ah ok, yeah this connects to your other comment. I was making an underlying assumption that a standalone register was going to always be continuous not a hybrid with standalone bits. This will do the wrong thing if a bit is in the circuit already with a standalone register. I'm not sure the best way to handle this, I guess I have to check all the indices in `bit_indices` list and detect a non-contiguous list and handle it accordingly

Reply:
I think the smoothest way forward is to merge the handling of standalone and shared registers (and have a follow up to fix the visualization). That is, always declare a circuit as being defined over a fixed number of `Qubit`s and `Clbit`s, and have named registers tag specific indices within those `Qubit` and `Clbit` lists. The "standalone" register is then just a special case where the bits are contiguous and happen to have been declared at the same time.